firestore/collgroupref.go

```diff
@@ -36,3 +45,107 @@ func newCollectionGroupRef(c *Client, dbPath, collectionID string) *CollectionGr
 		},
 	}
 }
+
+// GetPartitions returns a slice of QueryPartition objects. The number must be
+// positive. The actual number of partitions returned may be fewer.
+func (cgr CollectionGroupRef) GetPartitions(ctx context.Context, partitionCount int64) ([]QueryPartition, error) {
+	if partitionCount <= 0 {
+		return nil, errors.New("a positive partitionCount must be provided")
+	} else if partitionCount == 1 {
+		return []QueryPartition{{CollectionGroupQuery: cgr.Query, StartAt: "", EndBefore: ""}}, nil
+	}
+
+	db := cgr.c.path()
+	ctx = withResourceHeader(ctx, db)
+
+	// CollectiongGroup Queries need to be ordered by __name__ ASC
+	query, err := cgr.query().OrderBy(DocumentID, Asc).toProto()
+	if err != nil {
+		return nil, err
+	}
+	structuredQuery := &firestorepb.PartitionQueryRequest_StructuredQuery{
+		StructuredQuery: query,
+	}
+
+	// Uses default PageSize
+	pbr := &firestorepb.PartitionQueryRequest{
+		Parent:         db + "/documents",
+		PartitionCount: partitionCount,
+		QueryType:      structuredQuery,
+	}
+	cursorReferences := make([]*firestorepb.Value, 0, partitionCount)
+	iter := cgr.c.c.PartitionQuery(ctx, pbr)
+	for {
+		cursor, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		cursorReferences = append(cursorReferences, cursor.Values...)
+	}
+
+	// From Proto documentation:
+	// To obtain a complete result set ordered with respect to the results of the
+	// query supplied to PartitionQuery, the results sets should be merged:
+	// cursor A, cursor B, cursor M, cursor Q, cursor U, cursor W
+	// Once we have exhausted the pages, the cursor values need to be sorted in
+	// lexicographical order by segment (areas between '/').
+	sort.Sort(byReferenceValue(cursorReferences))
+
+	partitionQueries := make([]QueryPartition, 0, len(cursorReferences))
+	previousCursor := ""
+
+	for _, cursor := range cursorReferences {
+		cursorRef := cursor.GetReferenceValue()
+		qp := QueryPartition{
+			StartAt:   previousCursor,
+			EndBefore: cursorRef,
+		}
+		partitionQueries = append(partitionQueries, qp)
+		previousCursor = cursorRef
+	}
+
+	// In the case there were no partitions, we still add a single partition to
+	// the result, that covers the complete range.
+	lastPart := QueryPartition{
+		StartAt:   "",
+		EndBefore: "",
+	}
+	if len(cursorReferences) > 0 {
+		lastPart.StartAt = cursorReferences[len(cursorReferences)-1].GetReferenceValue()
+	}
+	partitionQueries = append(partitionQueries, lastPart)
+
+	return partitionQueries, nil
+}
+
+// byReferenceValue implements sort.Interface for []*firestorepb.Value
+
+type byReferenceValue []*firestorepb.Value
+
+func (a byReferenceValue) Len() int           { return len(a) }
+func (a byReferenceValue) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byReferenceValue) Less(i, j int) bool { return compareValues(a[i], a[j]) < 0 }
+
+// QueryPartition provides a Collection Group Reference and start and end split
```

Comment:
Again, it would be helpful to clarify what the user is actually supposed to use this for.